File: practice/algorithms/graph_theory/even_tree/EvenTree.scala
import scala.io.Source
import scala.collection.mutable.Map

object EvenTree extends App {
  val lines = Source.stdin.getLines().toList
  val count = lines.head.split(" ").map(_.toInt)
  val vertices = count(0)
  val edges = count(1)
  val edgesDesc = lines.tail

  val tree = Array.fill(vertices + 1)(0)
  val treeMap: Map[Int, Int] = Map.empty

  edgesDesc.foreach(desc => {
    val nums = desc.split(" ").map(_.toInt)
    val connected = nums(0)
    val parent = nums(1)
    tree(connected) = parent
    if(!treeMap.get(parent).isDefined) {
      treeMap.put(parent, 1)
    }
    if(!treeMap.get(connected).isDefined) {
      treeMap.put(connected, 1)
    }

    val count = treeMap.get(parent).get
    treeMap.put(parent, count + 1)
  })

  var result = 0
  for(nodeIndex <- Range(tree.length - 1, 1, -1)) {
    if(treeMap.get(nodeIndex).get % 2 == 0) {
      var hasEven = false
      for (i <- Range(tree.length - 1, 1, -1)) {
        if(tree(i) == nodeIndex) {
          val node = i
          if(treeMap.get(node).get % 2 == 0) {
            hasEven = true
          }
        }
      }
      if(!hasEven) {
        result += 1
        val parent = tree(nodeIndex)
        tree(nodeIndex) = 0
        val count = treeMap.get(parent).get
        treeMap.put(parent, count - 1)
      }
    }
  }
  println(result)
}

File: practice/algorithms/graph_theory/even_tree/EvenTree.java
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class EvenTree
{
	public static void main(String[] args)
	{
		Scanner scanner = new Scanner(System.in);
		int vertices = scanner.nextInt();
		int edges = scanner.nextInt();
		int[] tree = new int[vertices + 1];

		Map<Integer, Integer> map = new HashMap<>();

		for (int i = 0; i < edges; i++)
		{
			int connected = scanner.nextInt();
			int parent = scanner.nextInt();
			tree[connected] = parent;

			if(map.get(parent) == null)
			{
				map.put(parent, 1);
			}
			if(map.get(connected) == null)
			{
				map.put(connected, 1);
			}

			int count = map.get(parent);
			map.put(parent, ++count);
		}

		int result = 0;
		for (int nodeIndex = tree.length - 1; nodeIndex > 1; nodeIndex--)
		{
			if(map.get(nodeIndex) % 2 == 0)
			{
				boolean hasEven = false;
				for (int i = tree.length - 1; i > 1; i--)
				{
					if(tree[i] == nodeIndex)
					{
						int node = i;
						// System.out.println(node);
						if(map.get(node) % 2 == 0)
						{
							hasEven = true;
						}
					}
				}
				if(!hasEven)
				{
					result++;
					// System.out.println("cut " +nodeIndex);
					int parent = tree[nodeIndex];
					tree[nodeIndex] = 0;
					// System.out.println("parent " + parent);
					int count = map.get(parent);
					map.put(parent, --count);
				}
				// System.out.println();
			}
		}
		System.out.println(result);
		scanner.close();
	}
}
